### PR TITLE
final modifications for split polycommitment support in Snarky backend

### PR DESCRIPTION
**Pull request by ltvvlz**
This completes the integration of the split polynomial commitment schema for dlog-based Marlin protocol into the Snarky backend

The integration has been tested with a Snarky application implemented in double-marlin/concrete.ml

The unit test can be executed with the command:
dune runtest src/lib/double_marlin

### CODE REVIEW

File: src/lib/snarky_bn382_backend/dlog_based.ml [hunk 1]
```diff
@@ -141,7 +140,7 @@ module Keypair = struct
     in
     (set_urs_info, load)
 
-  let () = set_urs_info "/home/izzy/pickles-new/dlog-urs"
+  let () = set_urs_info "/home/pavel/urs/dlog-urs"
```
Comment (imeckler) [inline src/lib/snarky_bn382_backend/dlog_based.ml:143]: We should delete this or maybe make a file in `/tmp` or something

File: src/lib/snarky_bn382_backend/pairing_based.ml [hunk 2]
```diff
@@ -83,7 +103,7 @@ module Keypair = struct
   let set_urs_info, load_urs =
     let urs_info = Set_once.create () in
     let urs = ref None in
-    let set_urs_info ?(degree = 2 * 786_433) path =
+    let set_urs_info ?(degree = 3_11) path =
```
Comment (imeckler) [inline src/lib/snarky_bn382_backend/pairing_based.ml:106]: Is this number large enough

File: src/lib/snarky_bn382_backend/pairing_based.ml [hunk 3]
```diff
@@ -112,7 +132,7 @@ module Keypair = struct
     in
     (set_urs_info, load)
 
-  let () = set_urs_info "/home/izzy/pickles-new/pairing-urs"
+  let () = set_urs_info "/home/pavel/urs/pairing-urs"
```
Comment (imeckler) [inline src/lib/snarky_bn382_backend/pairing_based.ml:135]: same comment about the path

File: src/lib/double_marlin/concrete.ml [hunk 4]
```diff
@@ -0,0 +1,219 @@
+(*
```
[the remark this answers is not in the document]
Comment (ltvvlz) [reply src/lib/double_marlin/concrete.ml]: This has been addressed in commit 3e28ef7ec16d58af0d3c8ba087f5cad92d44907d

File: src/lib/double_marlin/concrete.ml [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+    let input () = Impl.Data_spec.[ Impl.Field.typ ]
+    let keys = Impl.generate_keypair ~exposing:(input ()) sqrt
+    let statement = Impl.Field.Constant.of_int 4
+    let proof = Impl.prove (Impl.Keypair.pk keys) (input ()) sqrt () statement
+    let () = assert (Impl.verify proof (Impl.Keypair.vk keys) (input ()) statement)
+end
+include SqrtExample(Snarky.Snark.Run.Make(Snarky.Snark.Backends.Bn128.Default) (Core.Unit))
+*)
+module RootExample (Impl : Snarky.Snark_intf.Run with type prover_state = unit) = struct
```
[the remark this answers is not in the document]
Comment (ltvvlz) [reply src/lib/double_marlin/concrete.ml]: This has been addressed in commit 3e28ef7ec16d58af0d3c8ba087f5cad92d44907d

File: src/lib/double_marlin/concrete.ml [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        Impl.assert_r1cs x x x2;
+        Impl.assert_r1cs x2 x2 x4;
+        Impl.assert_r1cs x4 x4 x16;
+        ()
+
+    let input () = Impl.Data_spec.[ Impl.Field.typ ]
+    let keys = Impl.generate_keypair ~exposing:(input ()) root
+    let statement = Impl.Field.Constant.of_int 256
+    let proof = Impl.prove (Impl.Keypair.pk keys) (input ()) root () statement
+    let () = assert (Impl.verify proof (Impl.Keypair.vk keys) (input ()) statement)
```
[the remark this answers is not in the document]
Comment (ltvvlz) [reply src/lib/double_marlin/concrete.ml]: This has been addressed in commit 3e28ef7ec16d58af0d3c8ba087f5cad92d44907d

File: src/lib/double_marlin/concrete.ml [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    let verified =
+    Snark.verify proof (Snark.Keypair.vk keypair) (input ()) root_start updates root_end
+
+    let () = Core.printf "Verified: %b\n" verified
+end
+
+include RootExample(Snarky.Snark.Run.Make(Snarky_bn382_backend.Pairing_based) (Core.Unit))
+include RootExample(Snarky.Snark.Run.Make(Snarky_bn382_backend.Dlog_based) (Core.Unit))
+include MerkleTree(Snarky.Snark.Make (Snarky_bn382_backend.Dlog_based))
+(*include MerkleTree(Snarky.Snark.Make (Snarky_bn382_backend.Pairing_based))*)
```
[the remark this answers is not in the document]
Comment (ltvvlz) [reply src/lib/double_marlin/concrete.ml]: This has been addressed in commit 3e28ef7ec16d58af0d3c8ba087f5cad92d44907d

File: src/lib/snarky_bn382_backend/dlog_based.ml (hunk 1 above)
Comment (ltvvlz) [reply src/lib/snarky_bn382_backend/dlog_based.ml]: This has been addressed in commit 3e28ef7ec16d58af0d3c8ba087f5cad92d44907d

File: src/lib/snarky_bn382_backend/pairing_based.ml (hunk 2 above)
Comment (ltvvlz) [reply src/lib/snarky_bn382_backend/pairing_based.ml:106]: Pairing backend is only used for a small circuit in the test. Have chosen small value for saving test time

File: src/lib/snarky_bn382_backend/pairing_based.ml (hunk 3 above)
Comment (ltvvlz) [reply src/lib/snarky_bn382_backend/pairing_based.ml]: This has been addressed in commit 3e28ef7ec16d58af0d3c8ba087f5cad92d44907d

**Comment by mrmr1993** [comment]
Closing, stale.

**Pull request closed without merging**